### PR TITLE
[SYCL-MLIR][NFC][LoopInternalization] Add an example

### PR DESCRIPTION
**Pull request by whitneywhtsang**
A pseudocode example is added to better understand the optimization for reader of the pass.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 1]
```diff
@@ -9,6 +9,49 @@
...
+//   %view = memref.view %shared_memory_ptr[%offset][%work_group_size]
+//                             : memref<-xi8, #sycl.access.address_space<local>>
+//                             to memref<?xf32,
+//                             #sycl.access.address_space<local>>
+//   // Loop is tiled with %work_group_size as the tile size.
+//   for (size_t k = 0; k < N; k+= work_group_size) {
+//     // Copy from global memory to shared local memory.
+//     // A's index needs to be adjusted to outer loop induction variable plus
+//     tiled loop lower bound, to get different portion of 'A'.
+//     %view[local_id] = A[local_id+k]
```
Comment (sommerlukas) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:42]: Shouldn't `A` be indexed with `global_id`?

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 2]
```diff
@@ -9,6 +9,49 @@
...
+//     // The optimization relies on each thread in a work group to initialize
+//     %view, so we need a barrier here before reading from %view.
+//     group_barrier(nditem.get_group());
+//     for (size_t t = k; t < k + work_group_size; ++t)
```
Comment (sommerlukas) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:46]: Does this loop need to be so complicated? Couldn't it simply iterate from `0` to `work_group_size`?

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 3]
```diff
@@ -9,6 +9,49 @@
...
 // local memory.
 //
+// Before optimization:
+//   size_t global_id = nditem.get_global_id(0)
+//   for (size_t k = 0; k < N; ++k)
+//     A[k];
+//
+// After optimization (pseudocode):
+//   // Obtain shared local memory for all candidate memory access in a kernel
+//   needed for this optimization.
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:19]:
```suggestion
//   // Obtain shared local memory for all candidate memory accesses in a kernel
//   needed for this optimization.
```

**Review by victor-eds — COMMENTED** [review]
No further comments apart from Lukas'

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 3 above)
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:19]: Remove double  '//'

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:46]: This is done by the loop tiling utility.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:19]: The double '//' is intentional to indicate that it is a comment in the example.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:42]: In this example, the original index of `A` is the loop induction variable `k`, after tiling, `A` is indexed by the tiled loop induction variable `t`, so we just need to adjust the index here by adding `local_id`.
If the original index of `A` is the `global_id`, then `A` should continue to be indexed with `global_id`.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp]: Changed `access` to `accesses`.

**Review by sommerlukas — APPROVED** [review]

**Review by etiotto — APPROVED** [review]

**Pull request merged**